service/history/execution/mutable_state_builder.go

```diff
@@ -2369,28 +2336,53 @@ func (e *mutableStateBuilder) logDuplicatedActivityEvents(events []*types.Histor
 				startedEventID:   event.ActivityTaskTimedOutEventAttributes.StartedEventID,
 			}
 		default:
-			return
+			return true
 		}
 
 		if _, ok := activityTaskUniqueEvents[uniqueEventParams]; ok {
-			e.logger.Error("Duplicate activity task event found",
+			e.logger.Warn("Duplicate activity task event found",
 				tag.WorkflowDomainName(e.GetDomainEntry().GetInfo().Name),
 				tag.WorkflowID(e.GetExecutionInfo().WorkflowID),
 				tag.WorkflowRunID(e.GetExecutionInfo().RunID),
 				tag.WorkflowScheduleID(scheduledEventID),
 				tag.WorkflowEventType(event.GetEventType().String()),
-				tag.Dynamic("duplication-source", duplicationSource),
 			)
 
 			e.metricsClient.IncCounter(metrics.HistoryFlushBufferedEventsScope, metrics.DuplicateActivityTaskEventCounter)
-		} else {
-			activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+			return false
 		}
+		activityTaskUniqueEvents[uniqueEventParams] = struct{}{}
+		return true
 	}
 
+	var headEvents []*types.HistoryEvent
+	var tailEvents []*types.HistoryEvent
+
+	// Sometimes we see buffered events are out of order when read back from database.  This is mostly not an issue
+	// except in the Activity case where ActivityStarted and ActivityCompleted gets out of order.  The following code
+	// is added to reorder buffered events to guarantee all activity completion events will always be processed at the end.
 	for _, event := range events {
-		checkActivityTaskEventUniqueness(event)
+		// We sometimes see duplicate events
+		if unique := checkActivityTaskEventUniqueness(event); !unique {
+			continue
+		}
+		switch event.GetEventType() {
+		case types.EventTypeActivityTaskCompleted,
+			types.EventTypeActivityTaskFailed,
+			types.EventTypeActivityTaskCanceled,
+			types.EventTypeActivityTaskTimedOut:
+			tailEvents = append(tailEvents, event)
+		case types.EventTypeChildWorkflowExecutionCompleted,
```

Comment:
@natemort @fimanishi dumb question: why are we doing this head / tail split?  I don't understand the intent

I kinda think this is a bugfarm having this list we're definately going to forget to update, could we put this attribute as part of the type? ie

```
// in the types package:

type HeadHistoryEventOrdering interface {
    IsHead bool
}

type EventTypeActivityTaskFailed int
func (EventTypeActivityTaskFailed) IsHead() {return true}

```

and here:
```
var headHistoryEntries []types.HistoryEvent
var tailHistoryEntries [][]types.HistoryEvent

for _, event := range events {
   if event.IsHead() {
      headHistoryEntries = append(headHistoryEntries, event)
   } else {
      tailHistoryEntries = append(tailHistoryEntries, event)
   }
}
```